### PR TITLE
Add support for wildcards in subdomains for valid redirect URIs

### PR DESCRIPTION
**Pull request by josh-cain**
See https://issues.jboss.org/browse/KEYCLOAK-3585.

This is a re-work of previous MR.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -121,6 +121,7 @@
         <hamcrest.version>1.3</hamcrest.version>
         <jmeter.version>2.10</jmeter.version>
         <junit.version>4.12</junit.version>
+        <mockito.version>1.10.19</mockito.version>
```
Comment (stianst) [inline pom.xml]: We have decided not to use mocking frameworks so we can't accept this

File: services/src/test/java/org/keycloak/protocol/oidc/utils/RedirectUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
+package org.keycloak.protocol.oidc.utils;
```
Comment (stianst) [inline services/src/test/java/org/keycloak/protocol/oidc/utils/RedirectUtilsTest.java]: Look at OAuthRedirectUriTest which can be extended to cover these tests properly

**Review by stianst — CHANGES_REQUESTED** [review]

**Comment by davidholsgrove** [comment]
Hi, any chance this PR will be merged into a Keycloak release soon?
Allowing the (optional) use of wildcard redirect urls on a controlled subdomain would vastly reduce administrative overhead for dynamic environment creation.
thanks,
David

**Comment by josh-cain** [comment]
Closing.  It's clear that this is not going to happen (at least not with this PR), and it's so stale at this point, might as well re-write it.

**Comment by sharedprophet** [comment]
@josh-cain Could you link the new PR here if you create one, so I can watch it?

**Comment by josh-cain** [comment]
@sharedprophet  - my apologies, I can see where that comment was not clear.  I have no intentions of re-writing this myself at this point.  Per @stianst 's comment, they're not going to take anything with mocks, so a re-write will entail one of:

 1. refactoring to decouple the data domain from a utility function as simple as wildcard validation (RealmModel), as you must have a properly populated RealmModel to test the exposed `verifyRedirectUri` function.
 1. manipulating the accessibility of the private `matchesRedirects` (commonly done with something like PowerMock, which I suppose would once again be taboo).
 1. writing massive integration tests involving an application server to test its ability to properly parse a url string since mocks cannot be used.

Pretty sure #1 or #2 above would get rejected, and I don't have the time/patience for #3.  Sorry to abandon it, hopefully the community can band together for this one soon!

**Pull request closed without merging**